fhir/com.b2international.snowowl.fhir.rest/src/com/b2international/snowowl/fhir/rest/FhirMediaType.java

```diff
@@ -0,0 +1,354 @@
+		}
+		
+		final String fhirVersionValue = mediaType.getParameter(MIME_TYPE_FHIR_VERSION_PARAMETER);
+		this.fhirVersion = FHIRVersion.fromCode(fhirVersionValue);
+	}
+
+	public MediaType getMediaType() {
+		return this.mediaType;
+	}
+	
+	public FhirFormat getFhirFormat() {
+		return fhirFormat;
+	}
+	
+	public FHIRVersion getFhirVersion() {
+		return fhirVersion;
+	}
+
+	public Resource parseResource(InputStream in) throws FHIRFormatError, IOException {
+		switch (fhirFormat) {
+		case JSON:
+			return parseResourceJson(in);
+		case XML:
+			return parseResourceXml(in); 
+		default: 
+			throw new NotImplementedException("No parser implementation found for format: " + fhirFormat);
+		}
+	}
+	
+	private Resource parseResourceJson(InputStream in) throws FHIRFormatError, IOException {
+		switch (fhirVersion) {
+		case _4_0:
+			org.hl7.fhir.r4.model.Resource r4 = new org.hl7.fhir.r4.formats.JsonParser().parse(in);
+			return VersionConvertorFactory_40_50.convertResource(r4);
+		case _4_3:
+			org.hl7.fhir.r4b.model.Resource r4b = new org.hl7.fhir.r4b.formats.JsonParser().parse(in);
+			return VersionConvertorFactory_43_50.convertResource(r4b);
+		case _5_0:
+			org.hl7.fhir.r5.model.Resource r5 = new org.hl7.fhir.r5.formats.JsonParser().parse(in);
+			return r5;
+		default: 
+			throw new NotImplementedException("No JSON parser implementation found for version: " + fhirVersion);
+		}
+	}
+	
+	private Resource parseResourceXml(InputStream in) throws FHIRFormatError, IOException {
+		switch (fhirVersion) {
+		case _4_0:
+			org.hl7.fhir.r4.model.Resource r4 = new org.hl7.fhir.r4.formats.XmlParser().parse(in);
+			return VersionConvertorFactory_40_50.convertResource(r4);
+		case _4_3:
+			org.hl7.fhir.r4b.model.Resource r4b = new org.hl7.fhir.r4b.formats.XmlParser().parse(in);
+			return VersionConvertorFactory_43_50.convertResource(r4b);
+		case _5_0:
+			org.hl7.fhir.r5.model.Resource r5 = new org.hl7.fhir.r5.formats.XmlParser().parse(in);
+			return r5;
+		default: 
+			throw new NotImplementedException("No XML parser implementation found for version: " + fhirVersion);
+		}
+	}
+
+	public void writeResource(ByteArrayOutputStream baos, Resource resource, boolean pretty) throws IOException {
+		switch (fhirFormat) {
+		case JSON:
+			writeResourceJson(baos, resource, pretty);
+			break;
+		case XML:
+			writeResourceXml(baos, resource, pretty);
+			break;
+		default: 
+			throw new NotImplementedException("No serializer implementation found for format: " + fhirFormat);
+		}
+	}
+	
+	private void writeResourceJson(ByteArrayOutputStream baos, Resource resource, boolean pretty) throws IOException {
+		switch (fhirVersion) {
+		case _4_0:
+			org.hl7.fhir.r4.model.Resource r4 = VersionConvertorFactory_40_50.convertResource(resource);
+			new org.hl7.fhir.r4.formats.JsonParser().setOutputStyle(pretty ? org.hl7.fhir.r4.formats.IParser.OutputStyle.PRETTY : org.hl7.fhir.r4.formats.IParser.OutputStyle.NORMAL).compose(baos, r4);
```

Comment:
Do `compose` calls close the output stream? (To be fair I'm not sure which one is needed, leaving it open until the end of the request or closing it when it is certain that no more output will be added.)

Reply:
ByteArrayOutputStreams does not need closing. See javadoc here:
https://docs.oracle.com/javase/8/docs/api/java/io/ByteArrayOutputStream.html

According to the format implementations the out stream is not closed after the compose method call completes. It's the responsibility of the caller, which is good I think.

Reply:
Aah, all right, I didn't see / remember that this does not go into the HTTP response directly (an `OutputStream` could be obtained for that as well) but it will get turned into an actual `byte[]` and that will be sent back in the form of a `ResponseEntity` with headers, status code, etc.